### PR TITLE
API Key resource

### PR DESCRIPTION
**Pull request by fantapop**
Added the API Key resource

This PR is based off the fantapop/service-account-resource branch which itself is not final and does not include a final version of the sdk.  I'll rebase these branches after the next prod deploy when we can get the latest version.

**Commit checklist**
- [x] Changelog
- [x] Doc gen (`make generate`)
- [x] Integration test(s)
- [x] Acceptance test(s)
- [x] Example(s)

- - -
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/terraform-provider-cockroach/203)

### CODE REVIEW

**Review by rgcase — CHANGES_REQUESTED** [review]
This one looks good too, just one comment about some commented code.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/terraform-provider-cockroach/203#-:-NwH9PR63FGHorrIb7ot:b-tlozw3)* status: 0 of 16 files reviewed, 2 unresolved discussions (waiting on @fantapop and @marksoper)

___
*[`docs/resources/api_key.md` line 56 at r1](https://reviewable.io/reviews/cockroachdb/terraform-provider-cockroach/203#-NwGc0MXEo5hHf8wn9mI:-NwGc0MXEo5hHf8wn9mJ:by55buo) ([raw file](https://github.com/cockroachdb/terraform-provider-cockroach/blob/5a1bcc4123753c6b8dbc947e0525e8084d78faae/docs/resources/api_key.md#L56)):*
> ```Markdown
> # during import.  The API key ID can be derived from the secret.
> # format: terraform import <resource> <api key secret>
> terraform import cockroach_api_key.example CCDB1_D4zMI3pZTmk5rGrzYqMhbc_NkcXLI8d81Mtx3djD45iwPfgtnaRv0XCh0Z9047K
> ```

Did you just generate this yourself somehow?



_Code quote:_
```Markdown
CCDB1_D4zMI3pZTmk5rGrzYqMhbc_NkcXLI8d81Mtx3djD45iwPfgtnaRv0XCh0Z9047K
```
___
*[`internal/provider/api_key_resource.go` line 147 at r1](https://reviewable.io/reviews/cockroachdb/terraform-provider-cockroach/203#-NwH7JOQ2AK__vaESbPU:-NwH7JOR5OcAWERwHTx9:b-z7ch46) ([raw file](https://github.com/cockroachdb/terraform-provider-cockroach/blob/5a1bcc4123753c6b8dbc947e0525e8084d78faae/internal/provider/api_key_resource.go#L147)):*
> ```Go
> 	apiKeyID := state.ID.ValueString()
> 
> 	// In case this was an import, validate the ID format and secret format
> ```

What do we want to do here? Delete the code or make it work?

**Review by fantapop — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/terraform-provider-cockroach/203)* status: 0 of 16 files reviewed, 2 unresolved discussions (waiting on @marksoper and @rgcase)

___
*[`docs/resources/api_key.md` line 56 at r1](https://reviewable.io/reviews/cockroachdb/terraform-provider-cockroach/203#-NwGc0MXEo5hHf8wn9mI:-NwHMsZ91_AJPTvekHeD:bs48rc9) ([raw file](https://github.com/cockroachdb/terraform-provider-cockroach/blob/5a1bcc4123753c6b8dbc947e0525e8084d78faae/docs/resources/api_key.md#L56)):*
<details><summary><i>Previously, rgcase (Ryan Case) wrote…</i></summary><blockquote>

Did you just generate this yourself somehow?
</blockquote></details>

This is from a key I generated on a local instance.
___
*[`internal/provider/api_key_resource.go` line 147 at r1](https://reviewable.io/reviews/cockroachdb/terraform-provider-cockroach/203#-NwH7JOQ2AK__vaESbPU:-NwHN5EvB2XVPCzQ5abH:b-1y1ahc) ([raw file](https://github.com/cockroachdb/terraform-provider-cockroach/blob/5a1bcc4123753c6b8dbc947e0525e8084d78faae/internal/provider/api_key_resource.go#L147)):*
<details><summary><i>Previously, rgcase (Ryan Case) wrote…</i></summary><blockquote>

What do we want to do here? Delete the code or make it work?
</blockquote></details>

oh thanks for pointing that out.  Let me test it now that its all working.

**Review by fantapop — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/terraform-provider-cockroach/203)* status: 0 of 16 files reviewed, 2 unresolved discussions (waiting on @marksoper and @rgcase)

___
*[`internal/provider/api_key_resource.go` line 147 at r1](https://reviewable.io/reviews/cockroachdb/terraform-provider-cockroach/203#-NwH7JOQ2AK__vaESbPU:-NwHOJnw1NO77uogHldu:bqzcdzi) ([raw file](https://github.com/cockroachdb/terraform-provider-cockroach/blob/5a1bcc4123753c6b8dbc947e0525e8084d78faae/internal/provider/api_key_resource.go#L147)):*
<details><summary><i>Previously, fantapop (Christopher Fitzner) wrote…</i></summary><blockquote>

oh thanks for pointing that out.  Let me test it now that its all working.
</blockquote></details>

I don't think we need further validate this here. To test, I modified the import ID in various ways.

1. changing the CCDB1 prefix, triggers and displays a good error as returned from the api:
```
│ Error: Error getting api key info
│
│ Unexpected error retrieving api key info: invalid argument: id is missing the expected prefix
```
2. changing the length of the id portion of the string, didn't trigger a specific validation on the length but I think the error it gave was reasonable:
```
│ Error: Cannot import non-existent remote object
│
│ While attempting to import an existing object to "cockroach_api_key.example3", the provider
│ detected that no object exists with the given id. Only pre-existing objects can be imported; check
│ that the id is correct and that it is associated with the provider's configured region or endpoint,
│ or use "terraform apply" to create a new remote object for this resource.
```

I didn't actually consider instantiating a client and testing the secret that was imported.  I could probably do that.  Do you think its reasonable or over the top?

**Review by fantapop — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/terraform-provider-cockroach/203)* status: 0 of 16 files reviewed, 2 unresolved discussions (waiting on @marksoper and @rgcase)

___
*[`internal/provider/api_key_resource.go` line 147 at r1](https://reviewable.io/reviews/cockroachdb/terraform-provider-cockroach/203#-NwH7JOQ2AK__vaESbPU:-NwHdLlA66RMGB7Ej5nG:b-m1fy6h) ([raw file](https://github.com/cockroachdb/terraform-provider-cockroach/blob/5a1bcc4123753c6b8dbc947e0525e8084d78faae/internal/provider/api_key_resource.go#L147)):*
<details><summary><i>Previously, fantapop (Christopher Fitzner) wrote…</i></summary><blockquote>

I don't think we need further validate this here. To test, I modified the import ID in various ways.

1. changing the CCDB1 prefix, triggers and displays a good error as returned from the api:
```
│ Error: Error getting api key info
│
│ Unexpected error retrieving api key info: invalid argument: id is missing the expected prefix
```
2. changing the length of the id portion of the string, didn't trigger a specific validation on the length but I think the error it gave was reasonable:
```
│ Error: Cannot import non-existent remote object
│
│ While attempting to import an existing object to "cockroach_api_key.example3", the provider
│ detected that no object exists with the given id. Only pre-existing objects can be imported; check
│ that the id is correct and that it is associated with the provider's configured region or endpoint,
│ or use "terraform apply" to create a new remote object for this resource.
```

I didn't actually consider instantiating a client and testing the secret that was imported.  I could probably do that.  Do you think its reasonable or over the top?
</blockquote></details>

I prototyped this but I don't have a good sense around whether it would actually be expected.  I guess it seems nice.

something like:

```
+       // Check to see if the API Key secret actually works
+       cl := newAPIService(ctx, req.ID)
+       _, _, err := cl.ListClusters(ctx, &client.ListClustersOptions{})
+       if err != nil {
+               resp.Diagnostics.AddError(
+                       "Invalid API Key",
+                       fmt.Sprintf(`Unexpected failure when testing the imported api key: %s`, formatAPIErrorMessage(err)))
+               return
+       }
```

example output:

```
│ Error: Invalid API Key
│
│ Unexpected failure when testing the imported api key: invalid api key
```

**Review by rgcase — APPROVED** [review]
:lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/terraform-provider-cockroach/203#-:-NwL7WRlE9zvCcw-J2V-:bg7igkv)* status: 0 of 16 files reviewed, 1 unresolved discussion (waiting on @fantapop and @marksoper)

___
*[`internal/provider/api_key_resource.go` line 147 at r1](https://reviewable.io/reviews/cockroachdb/terraform-provider-cockroach/203#-NwH7JOQ2AK__vaESbPU:-NwL77gV83ThqDeByHis:bckv3kj) ([raw file](https://github.com/cockroachdb/terraform-provider-cockroach/blob/5a1bcc4123753c6b8dbc947e0525e8084d78faae/internal/provider/api_key_resource.go#L147)):*
<details><summary><i>Previously, fantapop (Christopher Fitzner) wrote…</i></summary><blockquote>

I prototyped this but I don't have a good sense around whether it would actually be expected.  I guess it seems nice.

something like:

```
+       // Check to see if the API Key secret actually works
+       cl := newAPIService(ctx, req.ID)
+       _, _, err := cl.ListClusters(ctx, &client.ListClustersOptions{})
+       if err != nil {
+               resp.Diagnostics.AddError(
+                       "Invalid API Key",
+                       fmt.Sprintf(`Unexpected failure when testing the imported api key: %s`, formatAPIErrorMessage(err)))
+               return
+       }
```

example output:

```
│ Error: Invalid API Key
│
│ Unexpected failure when testing the imported api key: invalid api key
```
</blockquote></details>

Thanks for investigating this. I don't think you need to do this here, but I wonder if it makes sense for us to add a `ValidateAPiKey` request for things like this? I don't think you need it right now though.

Just deleting this commented code works for me.

File: examples/resources/cockroach_api_key/import.sh [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# Since the secret, is not retreivable after creation, it must be provided
+# during import.  The API key ID can be derived from the secret.
```
Comment (marksoper) [inline examples/resources/cockroach_api_key/import.sh:2]: Why/how is the API key ID derivable from the secret? @fantapop

Comment (fantapop) [reply examples/resources/cockroach_api_key/import.sh:2]: The secret contains the api key id as a prefix.  I don't know the reasons this was done but it makes it easy to determine if your secret matches a specific key ID which is handy.

**Pull request merged**